lxd/storage_volumes.go

```diff
@@ -1132,6 +1133,176 @@ func storagePoolVolumePost(d *Daemon, r *http.Request) response.Response {
+
+			sourceNodeOffline = sourceMemberInfo.IsOffline(s.GlobalConfig.OfflineThreshold())
+
+			return nil
+		})
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		var targetProject *api.Project
+		var targetMemberInfo *db.NodeInfo
+		var candidateMembers []db.NodeInfo
+
+		if sourceNodeOffline {
+			resp := forwardedResponseIfTargetIsRemote(s, r)
+			if resp != nil {
+				return resp
+			}
+
+			srcPool, err := storagePools.LoadByName(s, srcPoolName)
+			if err != nil {
+				return response.SmartError(err)
+			}
+
+			if srcPool.Driver().Info().Name == "ceph" {
+				// Load source volume.
+				srcPoolID, err := s.DB.Cluster.GetStoragePoolID(srcPoolName)
+				if err != nil {
+					return response.SmartError(err)
+				}
+
+				var dbVolume *db.StorageVolume
+
+				err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+					dbVolume, err = tx.GetStoragePoolVolume(ctx, srcPoolID, projectName, db.StoragePoolVolumeTypeCustom, volumeName, true)
+					return err
+				})
+				if err != nil {
+					// Check if the user provided an incorrect target query parameter and return a helpful error message.
+					_, volumeNotFound := api.StatusErrorMatch(err, http.StatusNotFound)
+					targetIsSet := r.URL.Query().Get("target") != ""
+					serverIsClustered, _ := lxdCluster.Enabled(s.DB.Node)
+
+					if serverIsClustered && targetIsSet && volumeNotFound {
+						return response.NotFound(fmt.Errorf("Storage volume not found on this cluster member"))
+					}
+
+					return response.SmartError(err)
+				}
+
+				req := api.StorageVolumePost{
+					Name: req.Name,
+				}
+
+				return storagePoolVolumeTypePostRename(s, r, srcPool.Name(), projectName, &dbVolume.StorageVolume, req)
+			}
+		} else {
+			resp := forwardedResponseToNode(s, r, req.Source.Location)
+			if resp != nil {
+				return resp
+			}
+		}
+
+		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+			p, err := dbCluster.GetProject(ctx, tx.Tx(), projectName)
+			if err != nil {
+				return err
+			}
+
+			targetProject, err = p.ToAPI(ctx, tx.Tx())
+			if err != nil {
+				return err
+			}
+
+			allMembers, err := tx.GetNodes(ctx)
+			if err != nil {
+				return fmt.Errorf("Failed getting cluster members: %w", err)
+			}
+
+			var targetGroupName string
+
+			targetMemberInfo, targetGroupName, err = project.CheckTarget(ctx, s.Authorizer, r, tx, targetProject, target, allMembers)
+			if err != nil {
+				return err
+			}
+
+			if targetMemberInfo == nil {
+				clusterGroupsAllowed := project.GetRestrictedClusterGroups(targetProject)
+
+				candidateMembers, err = tx.GetCandidateMembers(ctx, allMembers, nil, targetGroupName, clusterGroupsAllowed, s.GlobalConfig.OfflineThreshold())
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		})
+		if err != nil {
+			return response.SmartError(err)
+		}
+		// If no member was selected yet, pick the member with the least number of custom volumes.
```

Comment:
Does LXD have this existing logic in the client? If not then I dont think we need this here at this time.